06_gpu_and_ml/audio-editing/playdiffusion-api.py

```diff
@@ -123,6 +91,52 @@ def run_asr(audio):
             media_type="audio/wav"
         )
 
+@app.function(secrets=[modal.Secret.from_name("openai-secret")])
+def run_asr(audio_url):
+    audio_bytes, _ = write_to_tempfile(audio_url)
+
+    whisper_client = OpenAI(api_key=os.getenv("OPENAI_API_KEY"))
+    audio_file = open(audio_bytes, "rb")
+    transcript = whisper_client.audio.transcriptions.create(
+        file=audio_file,
+        model="whisper-1",
+        response_format="verbose_json",
+        timestamp_granularities=["word"]
+    )
+    word_times = [{
+        "word": word.word,
+        "start": word.start,
+        "end": word.end
+    } for word in transcript.words]
+
+    return transcript.text, word_times
+
+
+@app.local_entrypoint()
+def main(audio_url, output_text, output_path):
+    input_text, word_times = run_asr.remote(audio_url)
+    playdiffusion_model = Model()
+    output_audio = playdiffusion_model.generate(audio_url, input_text, output_text, word_times)
+    # Create output directory if it doesn't exist
+    os.makedirs("/tmp/playdiffusion", exist_ok=True)
+    
+    # Save the output audio to the specified path
+    output_path = os.path.join("/tmp/playdiffusion", "output.wav")
+    with open(output_path, "wb") as f:
+        f.write(output_audio)
+
+
+def write_to_tempfile(audio_url):
+    with tempfile.NamedTemporaryFile(delete=False, suffix='.wav') as temp_file:
+        # Download and write the audio to the temporary file
+        audio_bytes = urlopen(audio_url).read()
+        temp_file.write(audio_bytes)
+        temp_file.flush()  # Ensure data is written to disk
```

Comment:
I don't think there's a need to use `flush()` because this is a context manager that, when finished, it preserves this guarantee.